design/design-doc-composition-functions.md

```diff
@@ -175,604 +183,828 @@ metadata:
+  rpc RunFunction(RunFunctionRequest) returns (RunFunctionResponse) {}
+}
+
+// A RunFunctionRequest requests that a Composition Function be run.
+message RunFunctionRequest {
+  RequestMeta meta = 1;
+
+  // The observed state prior to invocation of a Function pipeline. State passed
+  // to each Function is fresh as of the time the pipeline was invoked, not as
+  // of the time each Function was invoked.
+  State observed = 2;
+
+  // Desired state according to a Function pipeline. The state passed to a
+  // particular Function may have been accumulated by processing a Composition's
+  // patch-and-transform resources array. It may also have been accumulated by
+  // previous Functions in the pipeline. Functions may mutate any part of the
+  // desired state they are concerned with, and must pass through any part of
+  // the desired state that they are not concerned with.
+  State desired = 3;
+
+  // Configuration specific to this Function invocation. A JSON representation
+  // of the 'config' block of the relevant entry in a Composition's 'functions'
+  // array.
+  optional google.protobuf.Struct config = 4;
+}
+
+// A RunFunctionResponse contains the result of a Composition Function run.
+message RunFunctionResponse {
+  ResponseMeta meta = 1;
+
+  // Desired state according to a Function pipeline. The state passed to a
+  // particular Function may have been accumulated by previous Functions in the
+  // pipeline. Functions may mutate any part of the desired state they are
+  // concerned with, and must pass through any part of the desired state that
+  // they are not concerned with.
+  State desired = 2;
+
+  // Results of the Function run. Results may be used for observability and
+  // debugging purposes.
+  repeated Result results = 3;
+}
+
+// RequestMeta contains metadata pertaining to a RunFunctionRequest.
+message RequestMeta {
+  // An opaque string identifying the content of the request. Two identical
+  // requests should have the same tag.
+  string tag = 1;
+}
+
+// ResponseMeta contains metadata pertaining to a RunFunctionResponse.
+message ResponseMeta {
+  // An opaque string identifying the content of the request.
+  string tag = 1;
+
+  // Time-to-live of this response. Idempotent functions with no side-effects
+  // (e.g. simple templating functions) may specify a TTL. Crossplane may choose
+  // to cache responses until the TTL expires.
+  optional google.protobuf.Duration ttl = 2;
+}
+
+// State of the composite resource (XR) and any composed resources.
+message State {
+  // The state of the composite resource (XR).
+  Resource composite = 1;
+
+  // The state of any composed resources.
+  map<string, Resource> resources = 2;
+}
+
+// A Resource represents the state of a resource.
+message Resource {
+  // The JSON representation of the resource.
+  google.protobuf.Struct resource = 1;
+
+  // The resource's connection details.
+  map<string, bytes> connection_details = 2;
+}
```

Comment:
Under the alpha design a Function can return a composed resource that specifies P&T style readiness checks and ways to derive connection details (e.g. `connectionSecretFromField` etc). I've removed that under this design. I don't think I've seen anyone use it, and I don't think we actually need it. We can always add it back if there's a demand.

It's possible for a Function to just derive and write whatever XR connection details it needs, and to set the XR's as Ready whenever it feels it is. I'm pretty sure @turkenh identified we didn't need this in the original design and I think I agreed but must have forgotten and built it anyway? See https://github.com/crossplane/crossplane/pull/2886#discussion_r991710579.